networkit/cpp/io/FastSNAPGraphReader.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * FastSNAPGraphReader.cpp
+ *
+ *  Created on: 04.05.2018
+ *      Author: Alexander van der Grinten
+ */
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "FastSNAPGraphReader.h"
+#include "../auxiliary/StringTools.h"
+#include "../auxiliary/Log.h"
+
+namespace NetworKit {
+
+Graph FastSNAPGraphReader::read(const std::string &path) {
+	Graph G;
+
+	// Maps SNAP node IDs to consecutive NetworKit node IDs.
+	auto mapNode = [&] (node in) -> node {
+		auto it = nodeIdMap.find(in);
+		if(it != nodeIdMap.end())
+			return it->second;
+		auto result = nodeIdMap.insert({in, G.addNode()});
+		assert(result.second);
+		return result.first->second;
+	};
+
+	// This function modifies the graph on input.
+	auto handleEdge = [&] (node source, node target) {
+		if(!G.hasEdge(source, target))
```

Comment:
This check can also be quite slow as it is a linear scan over all neighbors of that node. If memory does not matter, it is probably faster to store the edges in one large vector of node pairs and then sort this vector twice using bucket sort (or just parallel sorting). The graph could then be created using that vector.

Just a small but still significant improvement for undirected graphs is to execute hasEdge on the node with the smaller degree.

Reply:
Unfortunately I think memory does matter regarding I/O. Especially if the library is used on consumer machines with low RAM. 

Nevertheless the second idea can be very useful. Is there a reason this improvement is not already implemented in the ```Graph.cpp```?